graphql/core/validation/rules.py

```diff
@@ -591,7 +591,64 @@ def bad_value_for_default_arg_message(var_name, type, value):
 
 
 class VariablesInAllowedPosition(ValidationRule):
-    pass
+    visit_spread_fragments = True
+
+    def __init__(self, context):
+        super(VariablesInAllowedPosition, self).__init__(context)
+        self.var_def_map = {}
+        self.visited_fragment_names = set()
+
+    def enter_OperationDefinition(self, *args):
+        self.var_def_map = {}
+        self.visited_fragment_names = set()
+
+    def enter_VariableDefinition(self, node, *args):
+        self.var_def_map[node.variable.name.value] = node
+
+    def enter_Variable(self, node, *args):
+        var_name = node.name.value
+        var_def = self.var_def_map.get(var_name)
+
+        var_type = var_def and type_from_ast(self.context.get_schema(), var_def.type)
+        input_type = self.context.get_input_type()
+
+        if var_type and input_type and not self.var_type_allowed_for_type(self.effective_type(var_type, var_def),
+                                                                          input_type):
+            return GraphQLError(self.bad_var_pos_message(var_name, var_type, input_type),
+                                [node])
+
+    def enter_FragmentSpread(self, node, *args):
+        if node.name.value in self.visited_fragment_names:
+            return False
+
+        self.visited_fragment_names.add(node.name.value)
+
+    @staticmethod
+    def effective_type(var_type, var_def):
+        if not var_def.default_value or isinstance(var_def, GraphQLNonNull):
+            return var_type
+
+        return GraphQLNonNull(var_type)
+
+    @staticmethod
```

Comment:
How about changing this to `@classmethod` and shorten `VariablesInAllowedPosition` below to `cls`?